### DISCUSSION TITLE
Support isolating upstream groups in `dependsOn`

### DISCUSSION THREAD

**Discussion by GeorgeTaveras1231** [vercel/turborepo#2006] [Ideas]
First brought up in this question: https://github.com/vercel/turborepo/discussions/2005

Context
===
It is a common practice for library developers to create a distinction between **runtime** dependencies, and **build** dependencies by using `dependencies` (for runtime dependencies) and `devDependencies` (for **build** dependencies). 

Problem
===
At the time, there isn't a way to express that distinction in the `turbo.json` file's `dependsOn` config. 

The cost is that caches are unnecessarily purged in cases where a package's dependency changes, even though the dependency is not referenced during the build, and is only referenced during the runtime of the package.

#### package-1 > input
```js
import someDep from 'package-2';
console.log(someDep);
```
#### package-1 > output
```js
/* 
Maybe here some other things are added by the build, 
but `someDep` reference is intact because this is a runtime dependency
*/
import someDep from 'package-2';
console.log(someDep);
```

#### package-1/package.json

```json5
{
  "name": "package-1",
  "scripts": {
    "build": "some-internal-build-tool build" // << USE dev dependency
  },
  "dependencies": {
    "package-2": "^1.0.0"
  },
  "devDependencies": {
    "some-internal-build-tool": "^1.0.0"
  }
}
```

#### package-2/package.json

```json5
{
  "name": "package-2",
  "scripts": {
    "build": "some-internal-build-tool build" // << USE dev dependency
  },
  "devDependencies": {
    "some-internal-build-tool": "^1.0.0"
  }
}
```

#### some-internal-build-tool/package.json

```json5
{
  "name": "some-internal-build-tool",
  "scripts": {
    "build": "some-other-open-source-tool build" // << Uses external tool here
  },
  "devDependencies": {
    "some-other-open-source-tool ": "^1.0.0"
  }
}
```

#### turbo.json

```json5
{
  "$schema": "https://turborepo.org/schema.json",
  "pipeline": {
    "build": {
      "dependsOn": ["^build"],
      "outputs": ["lib/**"],
      "inputs": ["src/**"]
    }
  }
}
```

#### Then

If I make a change in `package-2`, I don't need to recompile `package-1` because this is a **runtime** dependency. However, if I make a change to `some-internal-build-tool` then I should need to recompile `package-1` and `package-2`.

Proposal (Idea)
===
Allow developers to isolate `dependencies` and `devDependencies` in the upstream reference. This will allow us to continue maintaining a distinction between **runtime** and **build** dependencies (Using `dependencies`, and `devDependencies` as proxies, respectively), while optimizing cache hits.

## Syntax

I don't know what the preference would be, but seems like it may make sense to just build on top of the convention to use an arbitrary symbol as a prefix (for upstream) and suffix (for downstream) [as `^dep` or `dep^` is used currently], but the symbol should be unique.

Additionally, it may make sense to suggest a more verbose but clear syntax to denote the dependency group reference

```json5
{
  "dependsOn": [
    {"input": "devDependencies", "target": "build"}
  ]
}
```

Final thoughts
===
I think allowing this distinction is a reasonable step forward in allowing developers more expressive freedom to optimize cache hits. 

However, I recognize that this is still limited and leads to unnecessary cache misses since there isn't a way to provide granular dependencies for a script. This a product of the `turbo.json`'s global nature, and the fact that `turbo` assumes that all dependencies (including `devDependencies`) can affect the output of a given command. 

A solution could be to allow providing granular dependencies per project/script (Like Bazel does with `BUILD` files).

For example, lets say 2 dependencies affect the result of `build`, and 2 other dependencies affect the result of `test`
```json5
{
  "scripts": {
    "build": "some-command",
    "test": "some-other-command"
  },
  // Maybe allow this in a package.json file
  "turbo": {
    "dependencies": {
      "build": [
        "pkg-1 build",
        "pkg-2 build"
      ],
      "test": [
        "pkg-3 build",
        "pkg-4 build"
      ]
    }
  },
  "devDependencies": {
    "pkg-1": "0.0.0",
    "pkg-2": "0.0.0"
    "pkg-3": "0.0.0",
    "pkg-4": "0.0.0",
  }
}
```

So my proposal is likely not the best solution, but a maybe reasonable step forward?

**Comment by nathanhammond** [comment]
There is some prior art in this space, but it's unclear to me how correct that can be. For example, a build-time dependency that does code injection or rewrites, or a in-repo runtime dependency that has a build step prior to being able to be consumed—neither of those can be classified singly so we can't simply introspect into `package.json` and infer information from where they're defined.

In _general_ it should be a safe assumption that if a user updated their `package.json` or lockfile they intended to affect _some_ change, but without solving the halting problem (impossible) we're unable truly guess where that change shows up, and which downstream tasks should be invalidated.

Focusing on "how do I tell my task graph about how this change affects the build?" is, eventually, an entire programming language. Trying to accomplish that in a purely declarative manner is going to be constraining and have rapidly diminishing returns.

We believe that the ideal behavior in this space is to build a task graph granular enough that the penalty for needing to process portions of the graph is negligible. That your task graph is human-understandable for manual configuration at a point early in a project is an aberration. There is a fantastic chart that demonstrates this which I can't find talking about Broccoli.js (an old JS build tool) which is basically two pictures:
- How I thought Broccoli would be used (a picture of 6 nodes in a graph, producing a single asset)
- How Broccoli is actually used (an export from a normal Broccoli build showing thousands of nodes, producing dozens of assets)

We don't disagree that this is an interesting space to investigate, but the solution for approaching this problem in a way that meets our exacting API standards and scales to ambitious projects has a very different shape.

**Comment by GeorgeTaveras1231** [comment]
@nathanhammond thanks for sharing your thoughts on this. I totally agree that finding a solution that allows (or comes closer to) 100% granularity can manifest as a whole new language. However, I feel like there is a middle ground between 100% granularity and the current state. My proposal only suggests a bit more granularity (the distinction between dependencies and devDependencies), not 100% granularity - It isn't a perfect solution, but I believe it provides some value, and I imagine it also isn't much more complex to maintain.
